torch/csrc/distributed/c10d/TraceUtils.h

```diff
@@ -657,9 +659,6 @@ struct NCCLTraceBuffer {
         entry->duration_ = duration.value();
       }
     }
-
-    entry->retired_ = true;
```

Comment:
We should be holding guard.lock() when we reach this line and we should have re-verified id matches event id.  

The proposed change looks like it should also be ok to me, but I don't understand why the issue can happen unless something is wrong with our locking.

Ok I see it now. What I said is only true if 'can compute duration' is true, which is only true if timing is enabled. The fix looks right to me.